pkg/bootstrap/docker/dockerhelper/client.go

```diff
@@ -0,0 +1,285 @@
+	ctx, cancel := defaultContext()
+	defer cancel()
+	response, err := c.client.ContainerExecCreate(ctx, container, config)
+	if ctxErr := contextError(ctx); ctxErr != nil {
+		return nil, ctxErr
+	}
+	if err != nil {
+		return nil, err
+	}
+	return &response, err
+}
+
+func (c *dockerClient) ContainerExecAttach(execID string, stdIn io.Reader, stdOut, stdErr io.Writer) error {
+	ctx, cancel := defaultContext()
+	defer cancel()
+	response, err := c.client.ContainerExecAttach(ctx, execID, types.ExecConfig{
+		AttachStdin:  stdIn != nil,
+		AttachStdout: true,
+		AttachStderr: true,
+	})
+	if ctxErr := contextError(ctx); ctxErr != nil {
+		return ctxErr
+	}
+	if err != nil {
+		return err
+	}
+	defer response.Close()
+	return holdHijackedConnection(stdIn, stdOut, stdErr, response)
+}
+
+func (c *dockerClient) ContainerExecInspect(execID string) (*types.ContainerExecInspect, error) {
+	ctx, cancel := defaultContext()
+	defer cancel()
+	response, err := c.client.ContainerExecInspect(ctx, execID)
+	if ctxErr := contextError(ctx); ctxErr != nil {
+		return nil, ctxErr
+	}
+	return &response, err
+}
+
+func (c *dockerClient) ImageInspectWithRaw(imageID string, getSize bool) (*types.ImageInspect, []byte, error) {
+	ctx, cancel := defaultContext()
+	defer cancel()
+	image, raw, err := c.client.ImageInspectWithRaw(ctx, imageID, getSize)
+	if ctxErr := contextError(ctx); ctxErr != nil {
+		return nil, nil, ctxErr
+	}
+	return &image, raw, err
+}
+
+func (c *dockerClient) ImagePull(ref string, options types.ImagePullOptions, writer io.Writer) error {
+	ctx := context.Background()
+	response, err := c.client.ImagePull(ctx, ref, options)
+	if err != nil {
+		return err
+	}
+	defer response.Close()
+	_, err = io.Copy(writer, response)
+	return err
+}
+
+func defaultContext() (context.Context, context.CancelFunc) {
+	return context.WithTimeout(context.Background(), defaultDockerOpTimeout)
+}
+
+func contextError(ctx context.Context) error {
+	if ctx.Err() == context.DeadlineExceeded {
+		return operationTimeout{err: ctx.Err()}
+	}
+	return ctx.Err()
+}
+
+// redirectResponseToOutputStream redirect the response stream to stdout and stderr.
+func redirectResponseToOutputStream(outputStream, errorStream io.Writer, resp io.Reader) error {
+	if outputStream == nil {
+		outputStream = ioutil.Discard
+	}
+	if errorStream == nil {
+		errorStream = ioutil.Discard
+	}
+	_, err := stdcopy.StdCopy(outputStream, errorStream, resp)
+	return err
+}
+
+// holdHijackedConnection holds the HijackedResponse, redirects the inputStream to the connection, and redirects the response
+// stream to stdout and stderr.
+func holdHijackedConnection(inputStream io.Reader, outputStream, errorStream io.Writer, resp types.HijackedResponse) error {
+	receiveStdout := make(chan error)
+	if outputStream != nil || errorStream != nil {
+		go func() {
+			receiveStdout <- redirectResponseToOutputStream(outputStream, errorStream, resp.Reader)
+		}()
+	}
+
+	sendStdin := make(chan error)
```

Comment:
buffered != non-blocking read.  As things stand, if the select reads a value from receiveStdout before reading a value from sendStdin, the function will return.  However in this case the inputStream goroutine will block forever trying to write a value to sendStdin, because it is unbuffered and no-one will read from it once the function has returned.